File: indigo_core/src/buffer.rs
use crate::{rope::RopeExt as _, selection::Selection};
use anyhow::Context as _;
use camino::Utf8PathBuf;
use ropey::Rope;
use std::{cmp::min, fs::File, io::BufReader, path::PathBuf};

#[derive(Clone, Debug, Default)]
pub struct Buffer {
    path: Utf8PathBuf,
    contents: Rope,
    selection: Selection,
    is_modified: bool,
    is_read_only: bool,
    vertical_scroll: usize,
    horizontal_scroll: usize,
}

impl Buffer {
    #[must_use]
    pub fn path(&self) -> &Utf8PathBuf {
        &self.path
    }

    #[must_use]
    pub fn contents(&self) -> &Rope {
        &self.contents
    }

    #[must_use]
    pub fn selection(&self) -> &Selection {
        &self.selection
    }

    #[must_use]
    pub fn is_modified(&self) -> bool {
        self.is_modified
    }

    #[must_use]
    pub fn is_read_only(&self) -> bool {
        self.is_read_only
    }

    #[must_use]
    pub fn vertical_scroll(&self) -> usize {
        self.vertical_scroll
    }

    #[must_use]
    pub fn horizontal_scroll(&self) -> usize {
        self.horizontal_scroll
    }

    pub fn open(path: Utf8PathBuf) -> anyhow::Result<Self> {
        let file = File::open(PathBuf::from(path.clone())).context("Failed to open file")?;

        let is_read_only = file
            .metadata()
            .context("Failed to get file metadata")?
            .permissions()
            .readonly();

        let rope =
            Rope::from_reader(BufReader::new(file)).context("Failed to read file into rope")?;

        Ok(Self {
            path,
            contents: rope,
            selection: Selection::default(),
            is_modified: false,
            is_read_only,
            vertical_scroll: 0,
            horizontal_scroll: 0,
        })
    }

    #[must_use]
    pub fn scroll_up(&self, distance: usize) -> Self {
        self.scroll_to_line(self.vertical_scroll.saturating_sub(distance))
    }

    pub fn scroll_up_mut(&mut self, distance: usize) {
        self.scroll_to_line_mut(self.vertical_scroll.saturating_sub(distance));
    }

    #[must_use]
    pub fn scroll_down(&self, distance: usize) -> Self {
        self.scroll_to_line(self.vertical_scroll + distance)
    }

    pub fn scroll_down_mut(&mut self, distance: usize) {
        self.scroll_to_line_mut(self.vertical_scroll + distance);
    }

    #[must_use]
    pub fn scroll_left(&self, distance: usize) -> Self {
        self.scroll_to_column(self.horizontal_scroll.saturating_sub(distance))
    }

    pub fn scroll_left_mut(&mut self, distance: usize) {
        self.scroll_to_column_mut(self.horizontal_scroll.saturating_sub(distance));
    }

    #[must_use]
    pub fn scroll_right(&self, distance: usize) -> Self {
        self.scroll_to_column(self.horizontal_scroll + distance)
    }

    pub fn scroll_right_mut(&mut self, distance: usize) {
        self.scroll_to_column_mut(self.horizontal_scroll + distance);
    }

    #[must_use]
    pub fn scroll_to_line(&self, line: usize) -> Self {
        let last_line = self.contents.len_lines_indigo().saturating_sub(1);
        Self {
            vertical_scroll: min(line, last_line),
            ..self.clone()
        }
    }

    pub fn scroll_to_line_mut(&mut self, line: usize) {
        let last_line = self.contents.len_lines_indigo().saturating_sub(1);
        self.vertical_scroll = min(line, last_line);
    }

    #[must_use]
    pub fn scroll_to_column(&self, column: usize) -> Self {
        Self {
            horizontal_scroll: column,
            ..self.clone()
        }
    }

    pub fn scroll_to_column_mut(&mut self, column: usize) {
        self.horizontal_scroll = column;
    }

    pub fn assert_valid(&self) {
        // TODO: `selection` is valid in `contents` rope

        // TODO: `vertical_scroll` doesn't go beyond the last line of the `contents` rope

        // TODO: Buffer validation

        self.selection.assert_valid();
    }
}

File: src/Indigo/Core/Buffer.hs
module Indigo.Core.Buffer
  ( Buffer (..)
  , InMemoryBuffer
  , FileBuffer

    -- * Query
  , contents
  , selection
  , isReadOnly
  , verticalScroll
  , horizontalScroll

    -- * Modify
  , scrollUp
  , scrollDown
  , scrollLeft
  , scrollRight
  , scrollToLine
  , scrollToColumn

    -- * Consume

    -- * Internal
  , isValid
  )
where

import Indigo.Core.Buffer.File (FileBuffer)
import Indigo.Core.Buffer.File qualified as FileBuffer
import Indigo.Core.Buffer.InMemory (InMemoryBuffer)
import Indigo.Core.Buffer.InMemory qualified as InMemoryBuffer
import Indigo.Core.Rope (ColumnIndex, LineIndex, Rope)
import Indigo.Core.Selection (Selection)

data Buffer
  = InMemory !InMemoryBuffer
  | File !FileBuffer

contents :: Buffer -> Rope
contents = \case
  InMemory buffer -> InMemoryBuffer.contents buffer
  File buffer -> FileBuffer.contents buffer

selection :: Buffer -> Selection
selection = \case
  InMemory buffer -> InMemoryBuffer.selection buffer
  File buffer -> FileBuffer.selection buffer

isReadOnly :: Buffer -> Bool
isReadOnly = \case
  InMemory buffer -> InMemoryBuffer.isReadOnly buffer
  File buffer -> FileBuffer.isReadOnly buffer

verticalScroll :: Buffer -> LineIndex
verticalScroll = \case
  InMemory buffer -> InMemoryBuffer.verticalScroll buffer
  File buffer -> FileBuffer.verticalScroll buffer

horizontalScroll :: Buffer -> ColumnIndex
horizontalScroll = \case
  InMemory buffer -> InMemoryBuffer.horizontalScroll buffer
  File buffer -> FileBuffer.horizontalScroll buffer

scrollUp :: Word -> Buffer -> Buffer
scrollUp distance = \case
  InMemory buffer -> InMemory $ InMemoryBuffer.scrollUp distance buffer
  File buffer -> File $ FileBuffer.scrollUp distance buffer

scrollDown :: Word -> Buffer -> Buffer
scrollDown distance = \case
  InMemory buffer -> InMemory $ InMemoryBuffer.scrollDown distance buffer
  File buffer -> File $ FileBuffer.scrollDown distance buffer

scrollLeft :: Word -> Buffer -> Buffer
scrollLeft distance = \case
  InMemory buffer -> InMemory $ InMemoryBuffer.scrollLeft distance buffer
  File buffer -> File $ FileBuffer.scrollLeft distance buffer

scrollRight :: Word -> Buffer -> Buffer
scrollRight distance = \case
  InMemory buffer -> InMemory $ InMemoryBuffer.scrollRight distance buffer
  File buffer -> File $ FileBuffer.scrollRight distance buffer

scrollToLine :: LineIndex -> Buffer -> Buffer
scrollToLine line = \case
  InMemory buffer -> InMemory $ InMemoryBuffer.scrollToLine line buffer
  File buffer -> File $ FileBuffer.scrollToLine line buffer

scrollToColumn :: ColumnIndex -> Buffer -> Buffer
scrollToColumn column = \case
  InMemory buffer -> InMemory $ InMemoryBuffer.scrollToColumn column buffer
  File buffer -> File $ FileBuffer.scrollToColumn column buffer

isValid :: Buffer -> Bool
isValid = \case
  InMemory buffer -> InMemoryBuffer.isValid buffer
  File buffer -> FileBuffer.isValid buffer
